File: Tests/Conformance/AppEntryPrototype05822.c
#include <r4os/r4os.h>
#include <string.h>

static R4XStartImport make_import(uint32_t group, const void *table) {
    R4XStartImport item = {0}; item.group_id = group; item.flags = R4XSTART_IMPORT_FLAG_GROUP_INTERFACE; item.table = (uint64_t)(uintptr_t)table; return item;
}
static R4XStartContext make_context(R4XStartImport *imports, uint32_t count, const char *args) {
    R4XStartContext ctx = {0};
    ctx.magic = R4XSTART_MAGIC; ctx.abi_major = R4XSTART_ABI_MAJOR; ctx.abi_minor = R4XSTART_ABI_MINOR; ctx.size = R4XSTART_CONTEXT_SIZE;
    ctx.flags = R4XSTART_FLAG_IMPORTS_VALID; ctx.imports = (uint64_t)(uintptr_t)imports; ctx.import_count = count;
    ctx.args = (uint64_t)(uintptr_t)args; ctx.args_len = args != 0 ? strlen(args) : 0u; return ctx;
}

int main(void) {
    R4XStartR4Sys first_sys = {R4XSTART_R4SYS_MAGIC, R4XSTART_R4SYS_VERSION, R4XSTART_R4SYS_SIZE, 0};
    R4XStartR4Sys second_sys = {R4XSTART_R4SYS_MAGIC, R4XSTART_R4SYS_VERSION, R4XSTART_R4SYS_SIZE, 0};
    R4XStartR4Desk desk = {R4XSTART_R4DESK_MAGIC, R4XSTART_R4DESK_VERSION, R4XSTART_R4DESK_SIZE, 0};
    R4XStartR4Draw draw = {R4XSTART_R4DRAW_MAGIC, R4XSTART_R4DRAW_VERSION, R4XSTART_R4DRAW_SIZE, 0};
    R4XStartImport console_imports[] = {make_import(R4L_GROUP_R4SYS, &first_sys)};
    R4XStartImport desktop_imports[] = {make_import(R4L_GROUP_R4SYS, &second_sys), make_import(R4L_GROUP_R4DESK, &desk), make_import(R4L_GROUP_R4DRAW, &draw)};
    R4XStartContext console_ctx = make_context(console_imports, 1u, "console");
    R4XStartContext desktop_ctx = make_context(desktop_imports, 3u, "desktop");
    R4App console, desktop;
    if (!r4_status_succeeded(r4_app_init(&console_ctx, R4_APP_PROFILE_CONSOLE, &console))) return 1;
    if (!r4_app_has_group(&console, R4L_GROUP_R4SYS) || r4_app_has_group(&console, R4L_GROUP_R4DESK)) return 2;
    if (r4_status_succeeded(r4_app_init(&console_ctx, R4_APP_PROFILE_DESKTOP, &desktop))) return 3;
    if (!r4_status_succeeded(r4_app_init(&desktop_ctx, R4_APP_PROFILE_DESKTOP, &desktop))) return 4;
    if (!r4_app_has_group(&desktop, R4L_GROUP_R4DESK) || !r4_app_has_group(&desktop, R4L_GROUP_R4DRAW)) return 5;
    if (console.system.table == desktop.system.table) return 6;
    uint64_t len = 0; const uint8_t *args = r4_app_args(&console, &len);
    if (args == 0 || len != 7u || memcmp(args, "console", 7u) != 0) return 7;
    if (R4_APP_PROFILE_DESKTOP_REQUIRED_GROUPS != ((1u << R4L_GROUP_R4SYS) | (1u << R4L_GROUP_R4DESK) | (1u << R4L_GROUP_R4DRAW))) return 8;
    if (r4_status_succeeded(r4_app_init(&console_ctx, (R4AppProfile)99, &desktop))) return 9;
    return 0;
}

File: Tests/Conformance/AppEntryPrototype05822.zig
const std = @import("std");
const r4os = @import("r4os");

fn importFor(group: r4os.abi.R4LGroup, table: usize) r4os.abi.R4XStartImport {
    return .{ .group_id = @intFromEnum(group), .flags = r4os.abi.r4xstart_import_flag_group_interface, .table = table };
}

fn context(imports: []const r4os.abi.R4XStartImport, args: []const u8) r4os.abi.R4XStartContext {
    return .{
        .flags = r4os.abi.r4xstart_flag_imports_valid,
        .args = if (args.len == 0) 0 else @intFromPtr(args.ptr),
        .args_len = args.len,
        .imports = @intFromPtr(imports.ptr),
        .import_count = @intCast(imports.len),
    };
}

test "profiles enforce required groups and expose missing optional groups" {
    var sys: r4os.abi.R4XStartR4Sys = .{};
    var desk: r4os.abi.R4XStartR4Desk = .{};
    var draw: r4os.abi.R4XStartR4Draw = .{};
    const console_imports = [_]r4os.abi.R4XStartImport{importFor(.r4sys, @intFromPtr(&sys))};
    const desktop_imports = [_]r4os.abi.R4XStartImport{
        importFor(.r4sys, @intFromPtr(&sys)),
        importFor(.r4desk, @intFromPtr(&desk)),
        importFor(.r4draw, @intFromPtr(&draw)),
    };
    var console_context = context(&console_imports, "console");
    var desktop_context = context(&desktop_imports, "desktop");
    var console = switch (r4os.App.init(&console_context, .console)) {
        .value => |app| app,
        .failure => return error.UnexpectedFailure,
    };
    try std.testing.expect(console.hasGroup(.r4sys));
    try std.testing.expect(!console.hasGroup(.r4desk));
    try std.testing.expectEqualStrings("console", console.args());
    try std.testing.expect(console.allocator() == null);
    switch (r4os.App.init(&console_context, .desktop)) {
        .failure => |failure| try std.testing.expectEqual(r4os.abi.err_no_group, failure.raw_code),
        .value => return error.ExpectedFailure,
    }
    var desktop_app = switch (r4os.App.init(&desktop_context, .desktop)) {
        .value => |app| app,
        .failure => return error.UnexpectedFailure,
    };
    try std.testing.expect(desktop_app.desktop() != null and desktop_app.drawing() != null);
    try std.testing.expectEqual(r4os.abi.R4XStartAppClass.gui, desktop_app.profileMeta().app_class);
}

test "two App objects own independent bundle values" {
    var first_table: r4os.abi.R4XStartR4Sys = .{};
    var second_table: r4os.abi.R4XStartR4Sys = .{};
    const first_imports = [_]r4os.abi.R4XStartImport{importFor(.r4sys, @intFromPtr(&first_table))};
    const second_imports = [_]r4os.abi.R4XStartImport{importFor(.r4sys, @intFromPtr(&second_table))};
    var first_context = context(&first_imports, "one");
    var second_context = context(&second_imports, "two");
    var first = switch (r4os.App.init(&first_context, .service)) {
        .value => |app| app,
        .failure => return error.UnexpectedFailure,
    };
    var second = switch (r4os.App.init(&second_context, .service)) {
        .value => |app| app,
        .failure => return error.UnexpectedFailure,
    };
    try std.testing.expect(first.bundle.sys.? != second.bundle.sys.?);
    try std.testing.expectEqualStrings("one", first.args());
    try std.testing.expectEqualStrings("two", second.args());
    try std.testing.expect(@hasDecl(r4os, "lowlevel"));
}
